rust-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,270 @@
+/* EcDh functions. For further documentation please refer to symcrypt.h */
+
+use crate::curve_type::*;
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+/// EcDhKey is a wrapper around symcrypt_sys::PSYMCRYPT_ECKEY, this is to let rust handle the free'ing of this pointer
+/// EcDhKey must be allocated after EcDhExpandedCurve and free'd before EcDhExpandedCurve is free'd
+struct EcDhKey {
+    inner: symcrypt_sys::PSYMCRYPT_ECKEY,
+    expanded_curve: EcDhExpandedCurve,
+}
+
+struct EcDhExpandedCurve(symcrypt_sys::PSYMCRYPT_ECURVE);
+
+impl Drop for EcDhExpandedCurve {
+    fn drop(&mut self) {
+        unsafe { symcrypt_sys::SymCryptEcurveFree(self.0) }
+    }
+}
+
+impl EcDhKey {
+    pub fn new(curve: CurveType) -> Result<Self, SymCryptError> {
+        unsafe {
+            // Allocating expanded_curve needed for key derivation. This pointer must be dropped after the key is dropped.
+            let curve_ptr = symcrypt_sys::SymCryptEcurveAllocate(convert_curve(curve), 0);
+            if curve_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+            // expanded_curve needs to be wrapped to properly free the curve in the case there is an error in EcDhKey initialization
+            let expanded_curve = EcDhExpandedCurve(curve_ptr);
+
+            // Key must be dropped before curve is dropped
+            let key_ptr = symcrypt_sys::SymCryptEckeyAllocate(curve_ptr);
+            if key_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+            let key = EcDhKey {
+                inner: key_ptr,
+                expanded_curve: expanded_curve,
```

Comment:
```suggestion
                expanded_curve,
```

This is more a style suggestion, so I'll leave it up to your team, but you can just say `expanded_curve` here and that's sugar for `expanded_curve: expanded_curve`.